File: modules/activiti-webapp-explorer2/src/main/java/org/activiti/explorer/ui/custom/ToolBar.java
/* Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.activiti.explorer.ui.custom;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.activiti.explorer.ui.custom.ToolbarEntry.ToolbarCommand;
import org.activiti.explorer.ui.mainlayout.ExplorerLayout;

import com.vaadin.ui.Alignment;
import com.vaadin.ui.Button;
import com.vaadin.ui.HorizontalLayout;
import com.vaadin.ui.Label;

/**
 * Container for holding {@link ToolbarEntry}s.
 * 
 * @author Frederik Heremans
 * @author Joram Barrez
 */
public class ToolBar extends HorizontalLayout {
  
  private static final long serialVersionUID = 7957488256766569264L;
  
  protected Map<String, ToolbarEntry> entryMap;
  protected ToolbarEntry currentEntry;
  protected List<Button> actionButtons;

  public ToolBar() {
    entryMap = new HashMap<String, ToolbarEntry>();
    actionButtons = new ArrayList<Button>();
    
    setWidth("100%");
    setHeight(36, UNITS_PIXELS);
    addStyleName(ExplorerLayout.STYLE_TOOLBAR);
    setSpacing(true);
    setMargin(false, true, false, true);
    
    // Add label to fill excess space
    Label spacer = new Label();
    spacer.setContentMode(Label.CONTENT_XHTML);
    spacer.setValue("&nbsp;");
    addComponent(spacer);
    setExpandRatio(spacer, 1.0f);
  }
  
  /**
   * Add a new entry to the tool bar.
   */
  public ToolbarEntry addToolbarEntry(String key, String title, ToolbarCommand command) {
    if(entryMap.containsKey(key)) {
      throw new IllegalArgumentException("Toolbar already contains entry for key: " + key);
    }

    ToolbarEntry entry = new ToolbarEntry(key, title);
    if(command != null) {
      entry.setCommand(command);
    }
    
    entryMap.put(key, entry);
    addEntryComponent(entry);
    return entry;
  }
  
  /**
   * Add a new entry, which displays a pop-up-list when clicked. Items of that list can be added
   * on returned {@link ToolbarPopupEntry} instance.
   */
  public ToolbarPopupEntry addPopupEntry(String key, String title) {
    if(entryMap.containsKey(key)) {
      throw new IllegalArgumentException("Toolbar already contains entry for key: " + key);
    }
    
    ToolbarPopupEntry entry = new ToolbarPopupEntry(key, title);
    entryMap.put(key, entry);
    addEntryComponent(entry);
    return entry;
  }
  
  /**
   * Add a button to the toolbar. The buttons are rendered on the right of the 
   * toolbar.
   */
  public void addButton(Button button) {
    button.addStyleName(ExplorerLayout.STYLE_TOOLBAR_BUTTON);
    
    actionButtons.add(button);
    // Button is added after the spacer
    addComponent(button);
    setComponentAlignment(button, Alignment.MIDDLE_RIGHT);
  }
  
  public void removeAllButtons() {
    for(Button b : actionButtons) {
      removeComponent(b);
    }
  }
  
  public long getCount(String key) {
    ToolbarEntry toolbarEntry = entryMap.get(key);
    if(toolbarEntry == null) {
      throw new IllegalArgumentException("Toolbar doesn't contain an entry for key: " + key);
    }
    return toolbarEntry.getCount();
  }
 
  /**
   * Update the count field on the entry with the given key.
   */
  public void setCount(String key, Long count) {
    ToolbarEntry toolbarEntry = entryMap.get(key);
    if(toolbarEntry == null) {
      throw new IllegalArgumentException("Toolbar doesn't contain an entry for key: " + key);
    }
    toolbarEntry.setCount(count);
  }
  
  /**
   * Gets the entry for the given key. Returns null when entry is not present for the given key.
   */
  public ToolbarEntry getEntry(String key) {
    return entryMap.get(key);
  }
  
  /**
   * Set the entry active with the given key. Active entries will
   * have alternative style applied to them.
   */
  public synchronized void setActiveEntry(String key) {
    if(currentEntry != null) {
      currentEntry.setActive(false);
    }
    
    currentEntry = entryMap.get(key);
    if(currentEntry != null) {
      currentEntry.setActive(true);
    }
  }
  
  protected void addEntryComponent(ToolbarEntry entry) {
    addComponent(entry, getComponentCount() - 1 - actionButtons.size());
    setComponentAlignment(entry, Alignment.MIDDLE_LEFT);
  }
  
}

File: modules/activiti-ui/activiti-app/src/main/webapp/editor/editor-app/configuration/toolbar.js
/* Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

var KISBPM = KISBPM || {};
KISBPM.TOOLBAR_CONFIG = {
    "items" : [
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.SAVE",
            "cssClass" : "editor-icon editor-icon-save",
            "action" : "KISBPM.TOOLBAR.ACTIONS.saveModel"
        },
        {
            "type" : "separator",
            "title" : "",
            "cssClass" : "toolbar-separator"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.CUT",
            "cssClass" : "editor-icon editor-icon-cut",
            "action" : "KISBPM.TOOLBAR.ACTIONS.cut",
            "enabled" : false,
            "enabledAction" : "element"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.COPY",
            "cssClass" : "editor-icon editor-icon-copy",
            "action" : "KISBPM.TOOLBAR.ACTIONS.copy",
            "enabled" : false,
            "enabledAction" : "element"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.PASTE",
            "cssClass" : "editor-icon editor-icon-paste",
            "action" : "KISBPM.TOOLBAR.ACTIONS.paste",
            "enabled" : false
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.DELETE",
            "cssClass" : "editor-icon editor-icon-delete",
            "action" : "KISBPM.TOOLBAR.ACTIONS.deleteItem",
            "enabled" : false,
            "enabledAction" : "element"
        },
        {
            "type" : "separator",
            "title" : "TOOLBAR.ACTION.SAVE",
            "cssClass" : "toolbar-separator"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.REDO",
            "cssClass" : "editor-icon editor-icon-redo",
            "action" : "KISBPM.TOOLBAR.ACTIONS.redo",
            "enabled" : false
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.UNDO",
            "cssClass" : "editor-icon editor-icon-undo",
            "action" : "KISBPM.TOOLBAR.ACTIONS.undo",
            "enabled" : false
        },
        {
            "type" : "separator",
            "title" : "TOOLBAR.ACTION.SAVE",
            "cssClass" : "toolbar-separator"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ALIGNVERTICAL",
            "cssClass" : "editor-icon editor-icon-align-vertical",
            "action" : "KISBPM.TOOLBAR.ACTIONS.alignVertical",
            "enabled" : false,
            "enabledAction" : "element",
            "disableInForm" : true,
            "minSelectionCount" : 2
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ALIGNHORIZONTAL",
            "cssClass" : "editor-icon editor-icon-align-horizontal",
            "action" : "KISBPM.TOOLBAR.ACTIONS.alignHorizontal",
            "enabledAction" : "element",
            "enabled" : false,
            "disableInForm" : true,
            "minSelectionCount" : 2
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.SAMESIZE",
            "cssClass" : "editor-icon editor-icon-same-size",
            "action" : "KISBPM.TOOLBAR.ACTIONS.sameSize",
            "enabledAction" : "element",
            "enabled" : false,
            "disableInForm" : true,
            "minSelectionCount" : 2
        },
        {
        	"type" : "separator",
        	"title" : "TOOLBAR.ACTION.SAVE",
        	"cssClass" : "toolbar-separator",
        	"disableInForm" : true
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ZOOMIN",
            "cssClass" : "editor-icon editor-icon-zoom-in",
            "action" : "KISBPM.TOOLBAR.ACTIONS.zoomIn"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ZOOMOUT",
            "cssClass" : "editor-icon editor-icon-zoom-out",
            "action" : "KISBPM.TOOLBAR.ACTIONS.zoomOut"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ZOOMACTUAL",
            "cssClass" : "editor-icon editor-icon-zoom-actual",
            "action" : "KISBPM.TOOLBAR.ACTIONS.zoomActual"
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.ZOOMFIT",
            "cssClass" : "editor-icon editor-icon-zoom-fit",
            "action" : "KISBPM.TOOLBAR.ACTIONS.zoomFit"
        },
        {
            "type" : "separator",
            "title" : "TOOLBAR.ACTION.SAVE",
            "cssClass" : "toolbar-separator",
            "disableInForm" : true
        },
    	{
            "type" : "button",
            "title" : "TOOLBAR.ACTION.BENDPOINT.ADD",
            "cssClass" : "editor-icon editor-icon-bendpoint-add",
            "action" : "KISBPM.TOOLBAR.ACTIONS.addBendPoint",
            "id" : "add-bendpoint-button",
            "disableInForm" : true
    	},
    	{
    	    "type" : "button",
    	    "title" : "TOOLBAR.ACTION.BENDPOINT.REMOVE",
    	    "cssClass" : "editor-icon editor-icon-bendpoint-remove",
    	    "action" : "KISBPM.TOOLBAR.ACTIONS.removeBendPoint",
    	    "id" : "remove-bendpoint-button",
    	    "disableInForm" : true
    	},
        {
            "type" : "separator",
            "title": "",
            "cssClass" : "toolbar-separator",
            "disableInForm" : true
        },
        {
            "type" : "button",
            "title" : "TOOLBAR.ACTION.HELP",
            "cssClass" : "glyphicon glyphicon-question-sign",
            "action" : "KISBPM.TOOLBAR.ACTIONS.help"
        }
    ],
    
    "secondaryItems" : []
};